docs/pallas/tpu/distributed.md

```diff
@@ -0,0 +1,1313 @@
+<!-- #endregion -->
+
+```python id="PyAGnWc9yI8T" executionInfo={"status": "ok", "timestamp": 1722904801801, "user_tz": 420, "elapsed": 1978, "user": {"displayName": "Justin Fu", "userId": "17543197034567316452"}} outputId="1d8229bd-cab5-495f-93e9-fff2e41db480"
+import jax
+from jax import lax
+from jax import numpy as jnp
+from jax.experimental import mesh_utils
+from jax.experimental import pallas as pl
+from jax.experimental import shard_map
+from jax.experimental.pallas import tpu as pltpu
+
+P = jax.sharding.PartitionSpec
+
+num_devices = jax.local_device_count()
+assert num_devices > 1, "Please run this colab with more than one device."
+assert "TPU" in jax.devices()[0].device_kind, "Please run this colab with TPU devices."
+print(f"Running with {num_devices} {jax.devices()[0].device_kind} devices.")
+```
+
+<!-- #region id="DySMGNByclMi" -->
+## TPU Topologies
+
+TPUs are typically deployed in pods of multiple devices connected via a high-bandwidth interchip interconnect (ICI) for communication within the pod that is much faster than a typical network connection. For example, the specifications sheet for a [TPU v5p](https://cloud.google.com/tpu/docs/v5p) states an ICI bandwidth of 4.8Tb/s per chip (for reference, TPU v5p also has 21Tb/s of *local* HBM bandwidth). The ICI allows us to implement fast and performant distributed kernels that require high-bandwidth communication within a pod, and use the datacenter network for parallelization over less bandwidth-intensive operations, such as data-parallelism over a batch dimension.
+
+TPUs pods are typically arranged in a 3D torus topology. The following graphic gives several examples of configurations of different sizes.
+
+![tpu_topologies](https://cloud.google.com/static/tpu/docs/images/v4-topologies.png)
+
+Flattened as a graph, the torus can be visualized as follows. Each edge (orange and black) is a bidirectional connection between two devices. You will commonly hear rings in conjunction with discussion about device toplogies - a key feature of a torus is that when taking one slice along an axis of the pod, such as the nodes `[(0,1), (1, 1), (2, 1), (3, 1)]` or `[(0, 1), (1, 1)]`, we have a ring of devices. This is a feature we can use to simplify communication patterns within the pod.
+
+![tpu_torus](https://cloud.google.com/static/tpu/docs/images/untwisted-tori.png)
+
+
+
+
+<!-- #endregion -->
+
+<!-- #region id="tPSYaiUk_nu0" -->
+## Remote Direct Memory Access (DMA) Model
+
+TPUs communicate via the remote DMA model, which is a restricted form of communication between devices. In traditional multi-threaded programming, different threads can typically read and write to arbitrary locations in shared memory. On TPUs, any buffers allocated to a kernel (in SMEM, VMEM, or HBM) are visible to all devices, and communication happens only in terms of writes from one device's local buffer to another device's buffer via a combined send-receive operation.
```

Comment:
The latter two sentences are confusing to me. They're set up as if to draw a contrast between traditional multi-threaded programming and TPUs, but instead of a contrast, "different threads can typically read and write to arbitrary locations in shared memory" actually sounds pretty similar "any buffers allocated to a kernel [...] are visible to all devices." What does "are visible" mean exactly?

Also, there are several subtly distinct terms mentioned together here: devices vs threads, buffers vs memories. I'm not sure if the intent is to use these interchangeably or to draw distinctions.

How about we instead say something like (bear with me, I may not have the technical details right, but that's why I'm trying to make them more clear here!):

> In traditional multi-core programming, different cores read and write to the same shared memory. Cores communicate using standard load and store instructions to that memory. In contrast, TPU cores do not share a memory space; each core has its own memory spaces, and there are distinct operations to send and receive data between them. For example, Core A and Core B might each execute an operation to communicate, with Core A sending data to an address in Core B's memory space and vice-versa. These direct memory-to-memory transfers are performed asynchronously by the memory system.

I included that second-to-last sentence in an attempt to capture what I think you meant by "visible": the program running on Core A can express an address in a Core B memory space, namely an address of the form `(device_id, dst_ref)` where the first component specifies that we're talking about Core B's memory. But I'm not sure that sentence is necessary.

WDYT?

Reply:
I rewrote the paragraph as this which should be more clear:

"TPUs communicate via a push-only model known as a remote direct memory access (RDMA). A TPU is allowed to issue copy instruction to push from a local buffer to any buffer on another device within the same pod that executes asynchronously from the main program thread. However, a TPU can only read data that is stored locally. This is in contrast to more traditional multi-core programming where it is possible to both read from and write values to a shared memory."

Reply:
Love it!